src/legacy/core_plugins/kibana/public/discover/np_ready/angular/discover_state.ts

```diff
@@ -0,0 +1,238 @@
+    set: (value: AppState | null) => {
+      if (value) {
+        previousAppState = appStateContainer.getState();
+        appStateContainer.set(value);
+      }
+    },
+  };
+
+  const { start, stop } = syncState({
+    storageKey: APP_STATE_URL_KEY,
+    stateContainer: appStateContainerModified,
+    stateStorage,
+  });
+
+  return {
+    kbnUrlStateStorage: stateStorage,
+    appStateContainer: appStateContainerModified,
+    startSync: start,
+    stopSync: stop,
+    setAppState: (newPartial: AppState) => setState(appStateContainerModified, newPartial),
+    replaceUrlAppState: async (newPartial: AppState = {}) => {
+      const state = { ...appStateContainer.getState(), ...newPartial };
+      await stateStorage.set(APP_STATE_URL_KEY, state, { replace: true });
+    },
+    getAppFilters: () => getFilters(appStateContainer.getState()),
+    resetInitialAppState: () => {
+      initialAppState = appStateContainer.getState();
+    },
+    getPreviousAppState: () => previousAppState,
+    flushToUrl: () => stateStorage.flush(),
+    isAppStateDirty: () => !isEqualState(initialAppState, appStateContainer.getState()),
+  };
+}
+
+/**
+ * Helper function to merge a given new state with the existing state and to set the given state
+ * container
+ */
+export function setState(stateContainer: ReduxLikeStateContainer<AppState>, newState: AppState) {
+  const oldState = stateContainer.getState();
+  const mergedState = { ...oldState, ...newState };
+  if (!isEqualState(oldState, mergedState)) {
+    if (mergedState.query) {
+      mergedState.query = migrateLegacyQuery(mergedState.query);
+    }
+    stateContainer.set(mergedState);
+  }
+}
+
+/**
+ * Helper function to compare 2 different filter states
+ */
+export function isEqualFilters(filtersA: Filter[], filtersB: Filter[]) {
+  if (!filtersA && !filtersB) {
+    return true;
+  } else if (!filtersA || !filtersB) {
+    return false;
+  }
+  return esFilters.compareFilters(filtersA, filtersB, esFilters.COMPARE_ALL_OPTIONS);
+}
+
+/**
+ * helper function to extract filters of the given state
+ * returns a state object without filters and an array of filters
+ */
+export function splitState(state: AppState = {}) {
+  const { filters = [], ...statePartial } = state;
+  return { filters, state: statePartial };
+}
+
+/**
+ * Helper function to compare 2 different state, is needed since comparing filters
+ * works differently
+ */
+export function isEqualState(stateA: AppState, stateB: AppState) {
+  if (!stateA && !stateB) {
+    return true;
+  } else if (!stateA || !stateB) {
+    return false;
+  }
+  const { filters: stateAFilters = [], ...stateAPartial } = stateA;
+  const { filters: stateBFilters = [], ...stateBPartial } = stateB;
+  return _.isEqual(stateAPartial, stateBPartial) && isEqualFilters(stateAFilters, stateBFilters);
+}
+
+/**
+ * Helper function to return array of filter object of a given state
+ */
+const getFilters = (state: AppState): Filter[] => {
```

Comment:
You use it from one place. Are you sure we need it? What about 
`state?. filters || [];`

Reply:
you're right, and it turn our that the function that needs `getFilters` is redundant in the meantime. removed